### PR TITLE
Devices Browser Pagination

### LINKED ISSUE #2379: No device list paging
Follow up to https://github.com/flowforge/security/issues/16

* Team Devices: Lack of paging causing device listing pages to slowdown when lots of devices exist
* Instance Devices: Lack of paging causing device listing pages to slowdown when lots of devices exist

_Originally posted by @Steve-Mcl in https://github.com/flowforge/security/issues/16#issuecomment-1614867214_

### PR DESCRIPTION
**Pull request by Pezmc**
## Description

Updates the device browser on the devices page, application page and instances page to all support filtering and search both client and server side.

Of note:
 - When there is only one page of results, search and filter are performed client side
 - When there is more than one page, search and filter are handled server side
 - The team/application/instance devices APIs all now use the same getter for consistency
 - The status bars at the top of the device browsers, always include all devices, regardless of current page or filter state
 - The approach is designed to be semi-generalisable to other endpoints
 - This also moves the alerting for update/edit/assign to only appear _after_ the API call completes successfully (previously eager)

This PR does _not_ implement server side sort, when there is more than one page of results, sort is disabled. This is due to complexities implementing sort with cursor based pagination. There is follow up work being considered to add this back.

_Pagination forced to 5 per page locally for screenshots and video:_

Screenshot 2023-08-30 at 22 38 43

https://github.com/flowforge/flowforge/assets/507155/725a5939-660d-4374-abc0-f0f11e3ad9cc

## Related Issue(s)

Fixes #2379

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/flowforge/flowforge/blob/main/CONTRIBUTING.md)
 - [ ] Suitable unit/system level tests have been added and they pass - **Not complete**
 - [ ] Documentation has been updated - **Not relevant**
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts

### CODE REVIEW

File: forge/db/models/Device.js [hunk 1]
```diff
@@ -275,23 +278,48 @@ module.exports = {
                         { model: M.ProjectSnapshot, as: 'activeSnapshot', attributes: ['id', 'hashid', 'name'] }
                     ]
 
-                    includes.push({
```
Comment (Pezmc) [inline forge/db/models/Device.js:278]: This made `includeInstanceApplication` entirely redundant

File: forge/db/models/Device.js [hunk 2]
```diff
@@ -275,23 +278,48 @@ module.exports = {
...
+                            where: buildPaginationSearchClause(pagination, where, ['Device.name', 'Device.type', 'Device.id'], {}, order),
+                            include: pagination.statusOnly ? (projectInclude.include.where ? [projectInclude] : []) : includes,
+                            order,
+                            limit: pagination.statusOnly ? null : limit
                         }),
-                        this.count({ where })
+                        this.count({ where, include: (projectInclude.include.where ? [projectInclude] : []) })
                     ])
+
+                    let nextCursors = []
```
Comment (Pezmc) [inline forge/db/models/Device.js:291]: Part of the unfinished support for sorted pagination with cursors

**Comment by Pezmc** [comment]
There will likely be more merge conflicts from https://github.com/flowforge/flowforge/pull/2670 and https://github.com/flowforge/flowforge/pull/2683, will rebase/merge after those are both in main.

**Comment by Steve-Mcl** [comment]
Small bug (may be existing TBF) but we should address this:

Currently, after un-assigning a device (from both instance & application) page before device can be re-assigned to something else.


https://github.com/flowforge/flowforge/assets/44235289/3760e160-dfc6-429f-98e2-bdacdbc7f97b

File: frontend/src/mixins/vue-timers.js
Comment (knolleary) [inline frontend/src/mixins/vue-timers.js:1]: I'm not very familiar with this mixin or where/how it is used. I cannot see any other references in our codebase to the `timer-tick` event, so I'll go along with this.

**Review by knolleary — APPROVED** [review]
All looks good.

**Pull request merged**